### ISSUE TITLE
State IO

### ISSUE THREAD

**Issue by hkershaw-brown** [NCAR/DART#359]
Notes for state IO

* read_variables is public and is used by mpas_atm/update_bc.f90
* out of date comments in state_vector_iod_mod.f90 (e.g reference to construct_file_name_in() which no longer exists)
    * read_model_time directly from model_mod rather than assim_model_mod
    * out of date comments on "dart format read"
* io_filenames_mod has a bunch of references to diagnostic files which no longer exist
* io_filanmes_mod is doing a bunch of netcdf variable checking 
   * state_structure_mod also has a mix of netcdf file info vs. state info
* stages_to_write is in state_vector_io_mod - more sense to have this info in io_fileaames_mod?
*  see #283 for round robin assumption

* write_augmented_state - why the restriction on  `! if there is only one timestep in your input file insert the mean and sd if requested`

**Comment by hkershaw-brown** [comment]
In `create_and_open_state_output`, only 'time'  (lower case) can be the unlimited dimension.

https://github.com/NCAR/DART/blob/cbe0d41e9ce190366b3cc5cc1433becf2ebc421c/assimilation_code/modules/io/direct_netcdf_mod.f90#L1643-L1656

If you have an unlimited dimension in the state_structure_mod, e.g. WRF has 'Time' this gets created as a limited dimension.

This means that your created netcdf files have a different netcdf dimension structure than your model files. 

`has_unlimited` is a property of the state_stucture%domain, but it is (can be) different between netcdf files because of create_and_open_state_output.  I think this is why dart_time_io_mod is querying the netcdf file rather than the state_structure (see comment above). 

There is also `domain%variable(ivar)%var_has_unlim` which is per variable, set but never used.

**Comment by hkershaw-brown** [comment]
This is a hard coded 0 however, there is a `integer, parameter :: SINGLE_IO_TASK_ID = 0` in the module.

https://github.com/NCAR/DART/blob/cc1da67976feb3b092ac9eae7388f6181ef33154/assimilation_code/modules/io/direct_netcdf_mod.f90#L431-L434

* I don't think this needs a broadcast, since if all tasks are in the initialize_single_file_io routine, all tasks have access to the local variable `local_model_mod_will_write_state_variables`
* double check pe vs task: since `initialize_single_file_io`(**ens_handle**, file_handle) should you be using **ens_handle**%my_pe rather than my_task_id for this routine? Maybe it does not matter. Similarly for `read_single_file`:
https://github.com/NCAR/DART/blob/cc1da67976feb3b092ac9eae7388f6181ef33154/assimilation_code/modules/io/direct_netcdf_mod.f90#L536-L538

**Comment by nancycollins** [comment]
we need to rename the file format that puts all the ensemble members plus inflation copies, mean, sd, etc in a single netcdf file (that dart dictates the format of). 

we called it "single file" but as you point out that's confusing for the situation where there is only 1 member involved. the code is going to expect to read a netcdf file with specific dimension names and variable names. 

other suggestions for this format?  combination file, combined file, dart format file, ???

**Comment by nancycollins** [comment]
yes, because even though there is always only 1 input and output file, this item toggles on and off the dart netcdf format file vs a model netcdf file. 

the namelist variable name is confusing because we called dart format files "single file" no matter how many members are in it.

**Comment by hkershaw-brown** [comment]
yeah I get it, just thinking about refactoring. Maybe it would be better to have the file describe itself as 'dart format'

**Comment by nancycollins** [comment]
if the dart file had a global attribute to indicate it was "dart format" that would be good -- make it more self-describing.  then maybe a namelist item wouldn't be needed.  that would be nice. 

but we'd have to think about how the code could use it. it might need to open the file, look for the attribute and then decide whether to use the state structure setup info from the model's static_init_model or the dart defined format to read the file.  i'm not clear if there is an order of things that works however.  

it would be nice if the i/o code could use the state structure for reading a dart format file but with all the members in a single netcdf variable i don't know if it can.

**Comment by hkershaw-brown** [comment]
`assert_restart_names_initialized` is printing the error message from  `assert_file_info_initialized`

https://github.com/NCAR/DART/blob/4dadee739df650ab5eca62fa7c29add3d4c52bed/assimilation_code/modules/io/io_filenames_mod.f90#L201-L233

**Comment by hkershaw-brown** [comment]
add_domain_blank is not blank (just a state vector) it has 3 dimensions: location, member, time. The are appropriate for lorenz_X style models, but not necessarily appropriate for anything else.

https://github.com/NCAR/DART/blob/77bb8c2d64ffb5a86ded071520476b6dc4dccc4c/assimilation_code/modules/io/state_structure_mod.f90#L412-L448

```
state%domain(dom_id)%unique_dim_names(1)  = 'location'
state%domain(dom_id)%unique_dim_names(2)  = 'member'
state%domain(dom_id)%unique_dim_names(3)  = 'time'
```

The 'member' dimension for the single file IO - again blending state_structure & IO (which are not the same, but currently are in the same state_structure).

**Comment by hkershaw-brown** [comment]
* diagnostic structure in the code but not ~~used at all~~ half used? https://github.com/NCAR/DART/blob/924182f29ce23f2a4e4c53aa57cc7d00fdccf374/assimilation_code/modules/io/state_structure_mod.f90#L131-L135
 Note this routines comment says is is depreciated 
 https://github.com/NCAR/DART/blob/f193ab47f8079445f3a58c848a2b54469c130a84/assimilation_code/modules/io/state_structure_mod.f90#L131-L135

 Then Used with a comment "may not be needed"

 Then domain = 1 is hardcoded anyway
 https://github.com/NCAR/DART/blob/f193ab47f8079445f3a58c848a2b54469c130a84/assimilation_code/modules/io/direct_netcdf_mod.f90#L2266-L2273

**Comment by hkershaw-brown** [comment]
missing_in_state is for  some ensemble members missing, dry land (all ens missing missing, e.g. POP) skates though.

![Screenshot 2024-09-05 at 4 52 28 PM](https://github.com/user-attachments/assets/f615249c-fa2e-4231-93d5-761b270ec807)

note check for missing in state in assim_tools is mucho expensive (30% of runtime)

**Comment by hkershaw-brown** [comment]
Note, public routines 

          get_index_start,            &
          get_index_end,              &
          get_sum_variables,          &
          get_sum_variables_below,

so some model_mods (maybe other code?) assuming that they can use offset in the state vector. 

https://github.com/NCAR/DART/blob/80502a8e030f5308f557a326819c18cb5079c494/assimilation_code/modules/io/state_structure_mod.f90#L111-L114
